### DISCUSSION TITLE
VirtualMixer usage

### DISCUSSION THREAD

**Discussion by kenavolic** [PlusToolkit/PlusLib#857] [Q&A]
Hi,

I am doing some experiments with the VirtualMixer. It works well for mixing video and optical tracking virtual device. I decided to add a third input that for now simulates a device that just sends a constant matrix:

```
PlusStatus vtkPlusGenericSensorTracker::InternalUpdate()
{
  double unfilteredTimestamp = vtkIGSIOAccurateTimer::GetSystemTime();

  // Get data
  if (this->AccelerometerTool != NULL)
  {
    vtkSmartPointer<vtkMatrix4x4> orientationSensorToTracker = vtkSmartPointer<vtkMatrix4x4>::New();
    // check if updated and retrieve timestamp and values
    orientationSensorToTracker->SetElement(0,0,0.5);

    // This device has no frame numbering, so just auto increment tool frame number
    unsigned long frameNumber = this->AccelerometerTool->GetFrameNumber() + 1 ;
    ToolTimeStampedUpdate(this->AccelerometerTool->GetId(), orientationSensorToTracker, TOOL_OK, frameNumber, unfilteredTimestamp);
    LOG_INFO("vtkPlusGenericSensorTracker::InternalUpdated from " << vtkIGSIOAccurateTimer::GetInstance() << " at " << unfilteredTimestamp);
  }
  return PLUS_SUCCESS;
}
```

The configuration file [PlusDeviceSet_Server_OpticalMarkerTrackerMixer.txt](https://github.com/PlusToolkit/PlusLib/files/7391225/PlusDeviceSet_Server_OpticalMarkerTrackerMixer.txt) sets up 2 servers one with the mixer, the other one with the generic device only. The server with the generic device only receives the right matrix but the server handling the virtual mixer always receives the identity matrix... Could it be a timestamp issue as we can see there is a difference in the frac part of the timestamp?

```
Matrix for AccelerometerToTrack TRANSFORM received in the server handling the generic device: 
=============
0.5, 0, 0, 0
0, 1, 0, 0
0, 0, 1, 0
0, 0, 0, 1
=============
TRANSFORM timestamp on 140104383330048 is 1634832160s 874.89ms


Matrix for AccelerometerToTrack TRANSFORM received in the server handling the mixer device: 
=============
1, 0, 0, 0
0, 1, 0, 0
0, 0, 1, 0
0, 0, 0, 1
=============
TRANSFORM timestamp on 140104806467328 is 1634832160s 844.156ms
IMAGE timestamp on 140104806467328 is 1634832160s 844.156ms
```

Thanks for your help!

**Comment by adamrankin** [comment]
Could you upload a log file of a sample run?

Plus does timestamp interpolation of transform sources, so it should interpolate to the nearest available video frame timestamp. If should be sending the constant transform in the transform-only server.

Edit: there is usually a warning if interpolation fails

**Reply by kenavolic** [reply]
From what I remember, there was no warning in the logs. I am currently developing a new camera device and then I'll get back to my experiments with the accelerometer data in the mixer and let you know.
